File: scala/space-age/src/main/scala/SpaceAge.scala
import SpaceAge.OrbitalPeriodScale._

import scala.math.BigDecimal.RoundingMode

object SpaceAge {
  val earthOrbitalPeriodPerSecond: Double = 31557600

  def onEarth(seconds: Double): Double = orbitalPeriodDuration(seconds, Earth)
  def onMercury(seconds: Double): Double = orbitalPeriodDuration(seconds, Mercury)
  def onVenus(seconds: Double): Double = orbitalPeriodDuration(seconds, Venus)
  def onMars(seconds: Double): Double = orbitalPeriodDuration(seconds, Mars)
  def onJupiter(seconds: Double): Double = orbitalPeriodDuration(seconds, Jupiter)
  def onSaturn(seconds: Double): Double = orbitalPeriodDuration(seconds, Saturn)
  def onUranus(seconds: Double): Double = orbitalPeriodDuration(seconds, Uranus)
  def onNeptune(seconds: Double): Double = orbitalPeriodDuration(seconds, Neptune)

  def orbitalPeriodDuration(seconds: Double, scale: OrbitalPeriodScale.Scale): Double = {
    BigDecimal((seconds / scale.scale) / earthOrbitalPeriodPerSecond)
      .setScale(2, RoundingMode.HALF_DOWN)
      .toDouble
  }

  object OrbitalPeriodScale {
    sealed abstract class Scale(val scale: Double)
    case object Earth extends Scale(1)
    case object Mercury extends Scale(0.2408467)
    case object Venus extends Scale(0.61519726)
    case object Mars extends Scale(1.8808158)
    case object Jupiter extends Scale(11.862615)
    case object Saturn extends Scale(29.447498)
    case object Uranus extends Scale(84.016846)
    case object Neptune extends Scale(164.79132)
  }
}

File: typescript/space-age/space-age.ts
export default class SpaceAge {
    private static readonly EarthYearInSeconds: number = 31557600
    private static readonly EarthScale: number = 1;
    private static readonly MercuryScale: number = 0.2408467
    private static readonly VenusScale: number = 0.61519726
    private static readonly MarsScale: number = 1.8808158
    private static readonly JupiterScale: number = 11.862615
    private static readonly SaturnScale: number = 29.447498
    private static readonly UranusScale: number = 84.016846
    private static readonly NeptuneScale: number = 164.79132
    seconds: number

    constructor(spaceAge: number) {
        this.seconds = spaceAge
    }

    onEarth(): number {
        return this.spaceAge(SpaceAge.EarthScale)
    }

    onMercury(): number {
        return this.spaceAge(SpaceAge.MercuryScale)
    }

    onVenus(): number {
        return this.spaceAge(SpaceAge.VenusScale)
    }

    onMars(): number {
        return this.spaceAge(SpaceAge.MarsScale)
    }

    onJupiter(): number {
        return this.spaceAge(SpaceAge.JupiterScale)
    }

    onSaturn(): number {
        return this.spaceAge(SpaceAge.SaturnScale)
    }

    onUranus(): number {
        return this.spaceAge(SpaceAge.UranusScale)
    }

    onNeptune(): number {
        return this.spaceAge(SpaceAge.NeptuneScale)
    }


    private spaceAge(scale: number): number {
        return Math.round((this.seconds / scale / SpaceAge.EarthYearInSeconds) * 100) / 100
    }
}